File: scripts/upload_videos.py
#!/usr/bin/env python3
"""
Upload videos to Firebase Storage

Usage:
1. Download service account key from Firebase Console:
   - Go to Project Settings -> Service Accounts
   - Click "Generate new private key"
   - Save as firebase-service-account.json in project root

2. Run: python3 scripts/upload_videos.py
"""

import os
import sys
from pathlib import Path

try:
    from google.cloud import storage
except ImportError:
    print("Installing google-cloud-storage...")
    os.system("pip3 install google-cloud-storage")
    from google.cloud import storage

# Config
BUCKET_NAME = "aptly-study-app.firebasestorage.app"
VIDEOS_DIR = Path(__file__).parent.parent / "public" / "videos" / "fsm"
DESTINATION_PREFIX = "videos/fsm"

VIDEOS = [
    "history-of-facebook.mp4",
    "instagram-audience.mp4",
    "snapchat-messaging.mp4",
    "social-media-policy.mp4",
    "channel-selection.mp4",
    "campaign-objectives.mp4",
    "campaign-budget.mp4",
]

def find_service_account():
    """Find service account file"""
    paths = [
        Path.cwd() / "firebase-service-account.json",
        Path.cwd() / "service-account.json",
        Path.home() / ".config" / "gcloud" / "application_default_credentials.json",
    ]

    # Check GOOGLE_APPLICATION_CREDENTIALS env var
    env_path = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")
    if env_path:
        paths.insert(0, Path(env_path))

    for path in paths:
        if path.exists():
            return str(path)

    return None

def upload_video(bucket, local_path, destination):
    """Upload a single video file"""
    blob = bucket.blob(destination)

    # Check if already exists
    if blob.exists():
        print(f"  ⚠️  Already exists, skipping: {destination}")
        blob.make_public()
        return blob.public_url

    print(f"  📤 Uploading {local_path.name} ({local_path.stat().st_size / 1024 / 1024:.1f} MB)...")

    blob.upload_from_filename(
        str(local_path),
        content_type="video/mp4",
    )

    # Make public
    blob.make_public()

    print(f"  ✅ Done: {blob.public_url}")
    return blob.public_url

def main():
    print("🚀 Firebase Storage Video Upload\n")

    # Find service account
    sa_path = find_service_account()

    if not sa_path:
        print("❌ Service account file not found!\n")
        print("To fix this:")
        print("1. Go to Firebase Console -> Project Settings -> Service Accounts")
        print("2. Click 'Generate new private key'")
        print("3. Save as 'firebase-service-account.json' in project root")
        print("\nOr set GOOGLE_APPLICATION_CREDENTIALS environment variable")
        sys.exit(1)

    print(f"Using credentials: {sa_path}\n")

    # Initialize client
    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = sa_path
    client = storage.Client()
    bucket = client.bucket(BUCKET_NAME)

    print(f"Bucket: {BUCKET_NAME}\n")

    # Upload videos
    results = []

    for video in VIDEOS:
        local_path = VIDEOS_DIR / video

        if not local_path.exists():
            print(f"  ⚠️  File not found: {video}")
            continue

        destination = f"{DESTINATION_PREFIX}/{video}"
        url = upload_video(bucket, local_path, destination)
        results.append((video, url))

    # Print summary
    print("\n" + "=" * 60)
    print("📋 UPLOAD COMPLETE")
    print("=" * 60)
    print("\nUpdate fsmCourse.ts videoUrl values with these URLs:\n")

    for video, url in results:
        name = video.replace(".mp4", "").replace("-", " ").title()
        print(f"// {name}")
        print(f"videoUrl: '{url}',\n")

if __name__ == "__main__":
    main()

File: scripts/uploadVideos.mjs
/**
 * Upload videos to Firebase Storage
 *
 * Usage: node scripts/uploadVideos.mjs
 */

import { initializeApp, cert } from 'firebase-admin/app';
import { getStorage } from 'firebase-admin/storage';
import * as fs from 'fs';
import * as path from 'path';

// Initialize Firebase Admin
const serviceAccountPath = process.env.GOOGLE_APPLICATION_CREDENTIALS ||
  path.join(process.cwd(), 'firebase-service-account.json');

if (!fs.existsSync(serviceAccountPath)) {
  console.error('❌ Service account file not found at:', serviceAccountPath);
  console.log('\nTo fix this:');
  console.log('1. Go to Firebase Console -> Project Settings -> Service Accounts');
  console.log('2. Click "Generate new private key"');
  console.log('3. Save as firebase-service-account.json in project root');
  process.exit(1);
}

const serviceAccount = JSON.parse(fs.readFileSync(serviceAccountPath, 'utf8'));

initializeApp({
  credential: cert(serviceAccount),
  storageBucket: 'aptly-study-app.firebasestorage.app',
});

const bucket = getStorage().bucket();

// Videos to upload
const videosDir = path.join(process.cwd(), 'public/videos/fsm');
const videos = [
  'history-of-facebook.mp4',
  'instagram-audience.mp4',
  'snapchat-messaging.mp4',
  'social-media-policy.mp4',
  'channel-selection.mp4',
  'campaign-objectives.mp4',
  'campaign-budget.mp4',
];

async function uploadVideo(filename) {
  const localPath = path.join(videosDir, filename);
  const destination = `videos/fsm/${filename}`;

  if (!fs.existsSync(localPath)) {
    console.log(`⚠️  Skipping ${filename} - file not found`);
    return null;
  }

  const stats = fs.statSync(localPath);
  const sizeMB = (stats.size / (1024 * 1024)).toFixed(1);

  console.log(`📤 Uploading ${filename} (${sizeMB} MB)...`);

  try {
    await bucket.upload(localPath, {
      destination,
      metadata: {
        contentType: 'video/mp4',
        cacheControl: 'public, max-age=31536000', // Cache for 1 year
      },
    });

    // Make file public
    const file = bucket.file(destination);
    await file.makePublic();

    const publicUrl = `https://storage.googleapis.com/${bucket.name}/${destination}`;
    console.log(`✅ Uploaded: ${publicUrl}`);

    return { filename, url: publicUrl };
  } catch (error) {
    console.error(`❌ Failed to upload ${filename}:`, error.message);
    return null;
  }
}

async function main() {
  console.log('🚀 Starting video upload to Firebase Storage\n');
  console.log(`Bucket: ${bucket.name}\n`);

  const results = [];

  for (const video of videos) {
    const result = await uploadVideo(video);
    if (result) {
      results.push(result);
    }
  }

  console.log('\n📋 Upload Summary:');
  console.log('==================');

  if (results.length === 0) {
    console.log('No videos uploaded.');
    return;
  }

  console.log('\nUpdate fsmCourse.ts with these URLs:\n');
  for (const { filename, url } of results) {
    const name = filename.replace('.mp4', '').replace(/-/g, ' ');
    console.log(`// ${name}`);
    console.log(`videoUrl: '${url}',\n`);
  }
}

main().catch(console.error);
